src/main/java/org/opentripplanner/routing/graph/GraphIndex.java

```diff
@@ -392,6 +466,104 @@ public StopAndDistance(Stop stop, int distance){
+        private Set<AgencyAndId> filterByStops;
+        private Set<AgencyAndId> filterByRoutes;
+        private Set<String> filterByBikeRentalStation;
+        private Set<String> seen = new HashSet<String>();
+
+        public PlaceFinderTraverseVisitor(List<Object> filterByTypes,
+                Set<AgencyAndId> filterByStops,
+                Set<AgencyAndId> filterByRoutes,
+                Set<String> filterByBikeRentalStations) {
+            this.filterByTypes = filterByTypes;
+            this.filterByStops = filterByStops;
+            this.filterByRoutes = filterByRoutes;
+            this.filterByBikeRentalStation = filterByBikeRentalStations;
+        }
+        @Override public void visitEdge(Edge edge, State state) {
+        }
+        @Override public void visitEnqueue(State state) {
+        }
+        @Override public void visitVertex(State state) {
+            Vertex vertex = state.getVertex();
+            int distance = (int)state.getElapsedTimeSeconds();
+            if (vertex instanceof TransitStop) {
+                Stop stop = ((TransitStop)vertex).getStop();
+                addResults(stop, distance);
+            } else if (vertex instanceof BikeRentalStationVertex) {
+                BikeRentalStation station = ((BikeRentalStationVertex)vertex).getStation();
+                addResults(station, distance);
+            }
+        }
+        private void addResults(BikeRentalStation station, int distance) {
+            boolean passesTypesFilter = filterByTypes == null;
+            if (filterByTypes != null) {
+                if (filterByTypes.contains("BICYCLE_RENT")) {
+                    passesTypesFilter = true;
+                }
+            }
+            boolean passesBikeRentalStationsFilter = filterByBikeRentalStation == null;
+            if (filterByBikeRentalStation != null) {
+                if (filterByBikeRentalStation.contains(station.id)) {
+                    passesBikeRentalStationsFilter = true;
+                }
+            }
+            if (passesTypesFilter && passesBikeRentalStationsFilter) {
+                PlaceAndDistance place = new PlaceAndDistance(station, distance);
+                placesFound.add(place);
+            }
+        }
+        private void addResults(Stop stop, int distance) {
+            if (filterByStops != null && !filterByStops.isEmpty() && !filterByStops.contains(stop.getId())) return;
+
+            List<TripPattern> patterns = patternsForStop.get(stop)
+                .stream()
+                .filter(pattern -> filterByTypes == null || filterByTypes.isEmpty() || filterByTypes.contains(pattern.mode))
+                .filter(pattern -> filterByRoutes == null || filterByRoutes.isEmpty() || filterByRoutes.contains(pattern.route.getId()))
+                .filter(pattern -> pattern.canBoard(pattern.getStopIndex(stop)))
+                .collect(toList());
+
+            for (TripPattern pattern : patterns) {
```

Comment:
I've thought about it too but maybe it's clearer now as there is the side-effect with the seen rows.